codalab/lib/upload_manager.py

```diff
@@ -0,0 +1,187 @@
+        if simplify_archive:
+            self._simplify_archive(dest_path)
+
+    def _unpack_fileobj(self, source_filename, source_fileobj, dest_path, simplify_archive):
+        zip_util.unpack(zip_util.get_archive_ext(source_filename), source_fileobj, dest_path)
+        if simplify_archive:
+            self._simplify_archive(dest_path)
+
+    def _simplify_archive(self, path):
+        if not os.path.isdir(path):
+            return
+
+        files = [f for f in os.listdir(path) if not self._ignore_file_in_archive(f)]
+        if len(files) == 1:
+            self._simplify_directory(path, files[0])
+
+    # ... ignore files that match any of these exactly
+    IGNORE_FILE_EXACT = ['.DS_Store', '__MACOSX']
+    # ... ignore files that match any of these patterns
+    IGNORE_FILE_PATTERNS = [re.compile(s) for s in ['^\._.*']]
+
+    def _ignore_file_in_archive(self, filename):
+        if filename in self.IGNORE_FILE_EXACT:
+            return True
+        return any([pattern.match(filename) for pattern in self.IGNORE_FILE_PATTERNS])
+
+    def _simplify_directory(self, path, child_path=None):
+        if child_path is None:
+            child_path = os.listdir(path)[0]
+
+        temp_path = path + crypt_util.get_random_string()
+        path_util.rename(path, temp_path)
+        child_path = os.path.join(temp_path, child_path)
+        path_util.rename(child_path, path)
+        path_util.remove(temp_path)
+
+    def update_metadata_and_save(self, bundle, new_bundle):
+        """
+        Updates the metadata about the contents of the bundle, including
+        data_size as well as the total amount of disk used by the user.
+
+        If |new_bundle| is True, saves the bundle as a new bundle. Otherwise,
+        updates it.
+        """
+        bundle_path = self._bundle_store.get_bundle_location(bundle.uuid)
+
+        dirs_and_files = None
+        if os.path.isdir(bundle_path):
+            dirs_and_files = path_util.recursive_ls(bundle_path)
+        else:
+            dirs_and_files = [], [bundle_path]
+
+        data_hash = '0x%s' % (path_util.hash_directory(bundle_path, dirs_and_files))
+        data_size = path_util.get_size(bundle_path, dirs_and_files)
+
+        if new_bundle:
+            bundle.data_hash = data_hash
+            bundle.metadata.set_metadata_key('data_size', data_size)
+            self._bundle_model.save_bundle(bundle)
+        else:
+            bundle_update = {
+               'data_hash': data_hash,
+               'metadata': {
+                    'data_size': data_size,             
+                },
+            }
+            self._bundle_model.update_bundle(bundle, bundle_update)
+
+        self._bundle_model.update_user_disk_used(bundle.owner_id)
+
+    def has_contents(self, bundle):
+        return os.path.exists(self._bundle_store.get_bundle_location(bundle.uuid))
+
+    def cleanup_existing_contents(self, bundle):
+        self._bundle_store.cleanup(bundle.uuid, dry_run=False)
```

Comment:
When is this called?  If we're deleting the bundle, then why update the bundle metadata?